src/main/scala/nl/knaw/dans/easy/fedoratobag/EasyFedora2vaultApp.scala

```diff
@@ -169,24 +192,61 @@ class EasyFedoraToBagApp(configuration: Configuration) extends DebugEnhancedLogg
     bag.addTagFile(content.serialize.inputStream, Paths.get(s"metadata/depositor-info/agreements.xml"))
   }
 
-  private def addPayloadFileTo(bag: DansV0Bag)(fedoraFileId: String): Try[Node] = {
-    val streamId = "EASY_FILE"
+  private def addPayloads(bag: DansV0Bag, fileFilterType: FileFilterType, fileIds: Seq[String]) = {
+    fileIds.toList.traverse(getFileInfo)
+      .flatMap {
+        fileInfos =>
+          val selected = fileFilterType match {
+            case FileFilterType.ALL => fileInfos
+            case t @ _ => selectFileByType(t, fileInfos)
+          }
+          if (selected.nonEmpty) selected.traverse(addPayloadFileTo(bag))
+          else Failure(NoPayloadFilesException())
+      }
+  }
+
+  private def selectFileByType(fileFilterType: FileFilterType, fileInfos: List[FileInfo]): List[FileInfo] = {
+    val openAccessFileInfos = fileInfos.filter(_.accessibleTo == "ANONYMOUS")
+    val selected = openAccessFileInfos.filter(_.mimeType.startsWith(
+      fileFilterType match {
+        case FileFilterType.PDF => "application/pdf"
+        case FileFilterType.IMAGE => "image/"
+      }))
+    val reversedSelection = openAccessFileInfos.filter(_.mimeType.startsWith(
+      fileFilterType match {
+        case FileFilterType.PDF => "image/" // Sic!
+        case FileFilterType.IMAGE => "application/pdf" // Sic!
+      }))
+    if (selected.nonEmpty) List(selected.maxBy(_.size))
+    else if (reversedSelection.nonEmpty) List(reversedSelection.maxBy(_.size))
+         else if (openAccessFileInfos.nonEmpty) List(openAccessFileInfos.maxBy(_.size))
+              else List.empty[FileInfo]
+  }
+
+  private def getFileInfo(fedoraFileId: String): Try[FileInfo] = {
     for {
       foXml <- fedoraProvider.loadFoXml(fedoraFileId)
-      path = Paths.get((foXml \\ "file-item-md" \\ "path").text)
-      _ = logger.info(s"Adding $fedoraFileId to $path")
+      path <- Try { Paths.get((foXml \\ "file-item-md" \\ "path").text) }
+      sizeString <- Try { (foXml \\ "file-item-md" \\ "size").text }
+      mimeType <- Try { (foXml \\ "file-item-md" \\ "mimeType").text }
+      accessibleTo <- Try { (foXml \\ "file-item-md" \\ "accessibleTo").text }
```

Comment:
More efficient and at least one level of missing elements properly catched
```suggestion
      fileMD <- FoXml.getFileMD(foXml)
      path <- Try { Paths.get((fileMD \\ "path").text) }
      sizeString <- Try { (fileMD \\ "size").text }
      mimeType <- Try { (fileMD \\ "mimeType").text }
      accessibleTo <- Try { (fileMD \\ "accessibleTo").text }
```